sdks/python/apache_beam/io/sdf_restriction_provider.py

```diff
@@ -0,0 +1,73 @@
+from apache_beam.io.restriction_trackers import OffsetRestrictionTracker
+from apache_beam.transforms.core import RestrictionProvider
+from apache_beam.transforms import ptransform
+from apache_beam.transforms import core
+
+__all__ = ['SDFBoundedSourceRestrictionProvider']
+
+class SDFBoundedSourceRestrictionProvider(RestrictionProvider):
+  def __init__(self, source, desired_chunk_size=None):
+    self._source = source
+    self._desired_chunk_size = desired_chunk_size
+
+  def initial_restriction(self, element):
+    # Get initial range_tracker from source
+    range_tracker = self._source.get_range_tracker(None, None)
+    return (range_tracker.start_position(), range_tracker.stop_position())
+
+  def create_tracker(self, restriction):
+    # Make an assumption here: OffsetRestrictionTracker works for all BoundedSource.
+    # Current limitation: OffsetRestrictionTracker can only track integer
+    # Possible TODO: have a another RestrictionTracker?
+    return OffsetRestrictionTracker(restriction[0], restriction[1])
+
+  def split(self, element, restriction):
+    # Invoke source.split to get initial splitting results.
+    source_bundles = self._source.split(self._desired_chunk_size)
+    for source_bundle in source_bundles:
+      yield (source_bundle.start_position, source_bundle.stop_position)
+
+  def restriction_size(self, element, restriction):
+    # More precisely, this function returns the weight, rather the actual size.
+    return restriction[1] - restriction[0]
+
+
+
+class SDFBoundedSourceWrapper(ptransform.PTransform):
```

Comment:
Can you use this PTransform instead of having a duplicate definition of the DoFn inside of iobase.py#876